src/dmd/tokens.d

```diff
@@ -451,6 +451,74 @@ extern (C++) struct Token
         Identifier ident;
     }
 
+    bool opEquals()(auto ref const Token rhs) const
+    {
+        if (value != rhs.value)
+            return false;
+
+        with (TOK) switch (value)
+        {
+            case int32Literal,
+                uns32Literal,
+                int64Literal,
+                uns64Literal,
+                charLiteral,
+                wcharLiteral,
+                dcharLiteral:
+                // unsvalue is used for signed integers and characters by the lexer
+                return unsvalue == rhs.unsvalue;
+
+            case int128Literal, uns128Literal:
+                // the lexer doesn't handle these
+                assert(false, "not implemented");
+
+            case float32Literal,
+                float64Literal,
+                float80Literal,
+                imaginary32Literal,
+                imaginary64Literal,
+                imaginary80Literal:
+                    return floatvalue == rhs.floatvalue;
```

Comment:
incorrectly fails for NaN values

Reply:
I know there’s a function, `isNaN`, in Phobos. Do we already have something like this in DMD or should I copy the one from Phobos?